File: genui/core/component.py
# -*- coding: utf-8 -*-
"""UI组件抽象层, 定义各种UI组件的抽象表示"""

from abc import ABC, abstractmethod
from typing import Any, Dict, List, Literal, Optional, Union
from pydantic import BaseModel, Field


class ComponentBase(BaseModel, ABC):
    """UI组件基类, 所有组件都继承此类"""

    id: str = Field(description="组件唯一标识符")
    type: str = Field(description="组件类型")

    class Config:
        """Pydantic配置"""
        extra = "forbid"


class Button(ComponentBase):
    """按钮组件"""

    type: Literal["button"] = "button"
    text: str = Field(description="按钮文本")
    on_click: Optional[str] = Field(
        default=None,
        description="点击事件处理函数名称"
    )
    enabled: bool = Field(default=True, description="是否启用")
    width: Optional[int] = Field(default=None, description="宽度(像素)")
    height: Optional[int] = Field(default=None, description="高度(像素)")


class TextInput(ComponentBase):
    """文本输入框组件"""

    type: Literal["text_input"] = "text_input"
    placeholder: str = Field(default="", description="占位符文本")
    default_value: str = Field(default="", description="默认值")
    multiline: bool = Field(default=False, description="是否多行输入")
    width: Optional[int] = Field(default=None, description="宽度(像素)")
    height: Optional[int] = Field(default=None, description="高度(像素)")


class Label(ComponentBase):
    """文本标签组件"""

    type: Literal["label"] = "label"
    text: str = Field(description="标签文本")
    font_size: int = Field(default=12, description="字体大小")
    bold: bool = Field(default=False, description="是否加粗")
    color: Optional[str] = Field(default=None, description="文本颜色(hex)")


class Container(ComponentBase):
    """容器组件, 可包含其他组件"""

    type: Literal["container"] = "container"
    layout: Literal["vertical", "horizontal", "grid"] = Field(
        default="vertical",
        description="布局方式: vertical(垂直), horizontal(水平), grid(网格)"
    )
    children: List["Component"] = Field(
        default_factory=list,
        description="子组件列表"
    )
    padding: int = Field(default=10, description="内边距(像素)")
    spacing: int = Field(default=5, description="组件间距(像素)")
    width: Optional[int] = Field(default=None, description="宽度(像素)")
    height: Optional[int] = Field(default=None, description="高度(像素)")


class Checkbox(ComponentBase):
    """复选框组件"""

    type: Literal["checkbox"] = "checkbox"
    label: str = Field(description="复选框标签")
    checked: bool = Field(default=False, description="是否选中")
    on_change: Optional[str] = Field(
        default=None,
        description="状态变化事件处理函数"
    )


class RadioGroup(ComponentBase):
    """单选按钮组组件"""

    type: Literal["radio_group"] = "radio_group"
    label: str = Field(description="单选组标签")
    options: List[str] = Field(description="选项列表")
    selected: Optional[str] = Field(default=None, description="当前选中项")
    on_change: Optional[str] = Field(
        default=None,
        description="选择变化事件处理函数"
    )


class Dropdown(ComponentBase):
    """下拉选择框组件"""

    type: Literal["dropdown"] = "dropdown"
    label: str = Field(default="", description="下拉框标签")
    options: List[str] = Field(description="选项列表")
    selected: Optional[str] = Field(default=None, description="当前选中项")
    width: Optional[int] = Field(default=None, description="宽度(像素)")


# 定义组件联合类型
Component = Union[
    Button,
    TextInput,
    Label,
    Container,
    Checkbox,
    RadioGroup,
    Dropdown
]

# 更新Container的forward reference
Container.model_rebuild()

File: src/core/component.ts
// -*- coding: utf-8 -*-
// UI组件抽象层, 定义各种UI组件的Zod schema

import { z } from 'zod';

// ============================================================
// 按钮组件 Schema
// ============================================================
export const ButtonSchema = z.object({
  id: z.string(),
  type: z.literal('button'),
  text: z.string(),
  on_click: z.string().optional(),
  enabled: z.boolean().default(true),
  width: z.number().optional(),
  height: z.number().optional(),
});

/** 按钮组件类型 */
export type Button = z.infer<typeof ButtonSchema>;

// ============================================================
// 文本输入框组件 Schema
// ============================================================
export const TextInputSchema = z.object({
  id: z.string(),
  type: z.literal('text_input'),
  placeholder: z.string().default(''),
  default_value: z.string().default(''),
  multiline: z.boolean().default(false),
  width: z.number().optional(),
  height: z.number().optional(),
});

/** 文本输入框组件类型 */
export type TextInput = z.infer<typeof TextInputSchema>;

// ============================================================
// 文本标签组件 Schema
// ============================================================
export const LabelSchema = z.object({
  id: z.string(),
  type: z.literal('label'),
  text: z.string(),
  font_size: z.number().default(12),
  bold: z.boolean().default(false),
  color: z.string().optional(),
});

/** 文本标签组件类型 */
export type Label = z.infer<typeof LabelSchema>;

// ============================================================
// 复选框组件 Schema
// ============================================================
export const CheckboxSchema = z.object({
  id: z.string(),
  type: z.literal('checkbox'),
  label: z.string(),
  checked: z.boolean().default(false),
  on_change: z.string().optional(),
});

/** 复选框组件类型 */
export type Checkbox = z.infer<typeof CheckboxSchema>;

// ============================================================
// 单选按钮组组件 Schema
// ============================================================
export const RadioGroupSchema = z.object({
  id: z.string(),
  type: z.literal('radio_group'),
  label: z.string(),
  options: z.array(z.string()),
  selected: z.string().optional(),
  on_change: z.string().optional(),
});

/** 单选按钮组组件类型 */
export type RadioGroup = z.infer<typeof RadioGroupSchema>;

// ============================================================
// 下拉选择框组件 Schema
// ============================================================
export const DropdownSchema = z.object({
  id: z.string(),
  type: z.literal('dropdown'),
  label: z.string().default(''),
  options: z.array(z.string()),
  selected: z.string().optional(),
  width: z.number().optional(),
});

/** 下拉选择框组件类型 */
export type Dropdown = z.infer<typeof DropdownSchema>;

// ============================================================
// 手动定义递归类型以避免循环引用
// ============================================================

/** 容器组件类型(手动定义以支持递归) */
export interface Container {
  id: string;
  type: 'container';
  layout: 'vertical' | 'horizontal' | 'grid';
  children: ComponentType[];
  padding: number;
  spacing: number;
  width?: number;
  height?: number;
}

/** 所有组件的联合类型 */
export type ComponentType =
  | Button
  | TextInput
  | Label
  | Container
  | Checkbox
  | RadioGroup
  | Dropdown;

// ============================================================
// 组件联合类型 Schema (含递归Container)
// ============================================================

/**
 * 组件联合类型 Schema, 使用discriminatedUnion按type字段区分.
 * Container的children字段使用z.lazy()实现递归引用.
 */
export const ComponentSchema: z.ZodType<ComponentType> = z.discriminatedUnion(
  'type',
  [
    ButtonSchema,
    TextInputSchema,
    LabelSchema,
    z.object({
      id: z.string(),
      type: z.literal('container'),
      layout: z.enum(['vertical', 'horizontal', 'grid']).default('vertical'),
      children: z
        .array(z.lazy((): z.ZodType<ComponentType> => ComponentSchema))
        .default([]),
      padding: z.number().default(10),
      spacing: z.number().default(5),
      width: z.number().optional(),
      height: z.number().optional(),
    }),
    CheckboxSchema,
    RadioGroupSchema,
    DropdownSchema,
  ],
);

// ============================================================
// 容器组件 Schema (独立导出, 使用ComponentSchema引用)
// ============================================================
export const ContainerSchema = z.object({
  id: z.string(),
  type: z.literal('container'),
  layout: z.enum(['vertical', 'horizontal', 'grid']).default('vertical'),
  children: z
    .array(z.lazy((): z.ZodType<ComponentType> => ComponentSchema))
    .default([]),
  padding: z.number().default(10),
  spacing: z.number().default(5),
  width: z.number().optional(),
  height: z.number().optional(),
});
